Impostors and Palettes

McAwesome123 opened the pull request.

McAwesome123 committed:
Impostor Chance Configuration

- Can now configure impostor chances
    - It is NOT linear
    - Heavily weighted towards 0
- Added placeholder palette dropdown
- Now compiles to .scr by default
    - Which is a great change to put in the same commit that mainly affects the .exe part of the screensaver

McAwesome123 committed:
An attempt to make palette choice work

(spoiler: it doesn't work)

McAwesome123 committed:
some changes idk

- SpritePalette renamed to PaletteGroup
- less static casts but not necessarily less fucked
- no copy-pasted code in spritecontrol.cpp
- AllPaletteName changed back to PaletteName
- reverted to .exe build

McAwesome123 committed:
Partial success

- Screensaver now partially works
    - Only when using All palettes
- Release changed back to .scr
- When opening the config dialog, it will now check whether pattern and palette are set to a correct option and reset them if they aren't

McAwesome123 commented:
Need to fix the crash from the color count being set too high when changing the palettes option to something other than all.

pierce-smith1 commented on configdialog.cpp:
```diff
@@ -45,8 +50,15 @@ BOOL ConfigDialog::command(WPARAM wparam, LPARAM lparam) {
 		case IDC_YONK_PATTERN: {
 			if (HIWORD(wparam) == CBN_SELENDOK) {
-				return combobox_changed((HWND) lparam);
+				return combobox_changed((HWND) lparam, IDC_YONK_PATTERN);
```
You might be surprised to hear but I have no problem with how this is being done.

McAwesome123 committed:
Quick fix for palette crashes

Probably far from ideal

pierce-smith1 commented on resources.h:
```diff
@@ -73,42 +100,129 @@ enum PaletteName {
+		PaletteName::yette,
+		PaletteName::zehal,
+	}},
+	{ PaletteGroup::All, {	// :lkfear:
```
Let's do whatever we can to avoid repeating ourselves here since this is just asking for trouble. Here's an idea: since these palette names are just the numbers from 0 to 47 or whatever, let's just generate it at run time with a function:

```c++
inline std::vector<PaletteName> make_all_palettes() {
    std::vector<PaletteName> all_palettes((size_t) PaletteName::_PALETTE_COUNT);
    std::generate(all_palettes.begin(), all_palettes.end(), [n = 0]() mutable { return (PaletteName) n++; });
    return all_palettes;
}

// ...
{ PaletteGroup::All, make_all_palettes() }
```

This is some of the weirdest fucking c++ I've ever seen but it does seem to work. I stole it from [here](https://en.cppreference.com/w/cpp/algorithm/generate).

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -9,12 +9,13 @@ SpriteGenerator::SpriteGenerator() {
 
-	std::vector<int> bag_of_palettes(_PALETTE_COUNT);
-	std::iota(bag_of_palettes.begin(), bag_of_palettes.end(), 0);
+	//std::vector<int> bag_of_palettes(static_cast<int>(PaletteName::_PALETTE_COUNT));
```
Be sure to clean these comments up.

pierce-smith1 commented on configdialog.cpp:
```diff
@@ -92,8 +111,23 @@ void ConfigDialog::refresh() {
 		SendMessage(slider, TBM_SETPOS, TRUE, encodef(m_current_config.at(entry.first)));
 	}
 
+	if (m_current_config.at(YonkPattern) < 0 || m_current_config.at(YonkPattern) >= _PATTERN_COUNT)
```
Something about all of this is hurting my 3 am brain and I don't know what to make of it. I'll come back to this.

pierce-smith1 replied:
OK, here are some thoughts. First, I don't think we should show a message box in these cases, we should just silently fix the value and move on. I have a lot to say about this but really it just boils down to "I don't want to annoy users with a message box over something this unimportant."

Second, it feels like this validation + repair stuff really wants to be a separate system in the code. It might be as simple as just having a function `is_config_value_in_range` that we can call rather than doing these if checks manually. I want to think about this some more though, because I'm not 100% sure how we should design it, so I'll make another issue for it and we can address it later.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -2,19 +2,23 @@
+	std::vector<PaletteName> bag_of_palettes = PALETTES_BY_GROUP.at((PaletteGroup) (cfg.at(YonkPalette)));
 
-	for (int i = 0; i < round(cfg.at(MaxColors)); i++) {
+	float maxColors = cfg.at(MaxColors) * (bag_of_palettes.size() / config_ranges.at(MaxColors).second);
```
We really want `MaxColors` to be a value between 0.0 and 1.0, but we can't just do that because people's existing configs will be outside of that range. So I guess this is fine for now.

pierce-smith1 replied:
However - styling!! `snake_case` for locals please. I'm pointing this out because VS won't enforce it, even with the new .editorconfig.

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -2,19 +2,23 @@
-	for (int i = 0; i < round(cfg.at(MaxColors)); i++) {
+	float maxColors = cfg.at(MaxColors) * (bag_of_palettes.size() / config_ranges.at(MaxColors).second);
+
+	for (int i = 0; i < round(maxColors); i++) {
```
I'm scared of this. If the result of `round(maxColors)` is greater than `bag_of_palettes.size()`, the program will try to erase from an empty list and invoke undefined behavior. We cannot mathematically reason that this will never be the case because of floating-point imprecision. 

Since it's so important for this to not exceed `bag_of_palettes.size()`, I think we should do something like this, even though it is pretty clunky:

```c++
if (round(maxColors) > bag_of_palettes.size()) {
    maxColors = bag_of_palettes.size();
}
```

pierce-smith1 commented on spritecontrol.cpp:
```diff
@@ -24,7 +28,7 @@ std::vector<Sprite *> SpriteGenerator::make(unsigned int n) const {
 	for (float y = -1.2f; y < 1.2f; y += 1.0f / sqrt(cfg.at(SpriteCount))) {
 		for (float x = -1.2f; x < 1.2f; x += 1.0f / sqrt(cfg.at(SpriteCount))) {
-			if (Noise::random() < 0.002f) {
+			if (Noise::random() < pow(cfg.at(ImpostorChance), 3)) {
```
I may have said it already, but the cubic is pretty clever here.

pierce-smith1 commented on configdialog.cpp:
```diff
@@ -92,8 +111,23 @@ void ConfigDialog::refresh() {
+		(SpritePattern)m_current_config.at(YonkPattern)).c_str()
+	);
+
+	if (m_current_config.at(YonkPalette) < 0 || m_current_config.at(YonkPalette) >= static_cast<int>(PaletteGroup::_PALETTE_OPTION_COUNT))
```
Why does there need to be a `static_cast` here, but not on the previous if? And as a style point (that VS is not set up to enforce), we should use C style casts instead of `static_cast`.

McAwesome123 replied:
Wasn't aware that a C-style cast works here. And it's needed because the compiler doesn't like comparing an int to an enum class member.

pierce-smith1 commented on Resource.rc:
![image](https://github.com/pierce-smith1/yok.scr/assets/50504215/36f11041-fcbd-4d8b-9fe7-888b6c3ce5d5)

Two comments on the dialog: 
* The label "Palette" should maybe be pluralized into "Palettes," since we don't want people getting confused and thinking they're selecting one palette for all the yokers.
* The whole "Family" box a just a touch too tall, with a little bit too much empty space on the bottom. It could be shrunk just a tiny bit.

pierce-smith1 reviewed (changes requested):
I think this got everything out of my system. I'm going to wait until this PR gets merged before merging any of my other ones, because I want to impending apocalyptic merge conflict shitstorm to fall on me, not you.

McAwesome123 committed:
Clean-up and fixes

- All palettes enum is now run-time generated rather than copy-pasted
- Comment clean-up and minor code shifting
- Fixed rounding/out-of-range errors for max colors?
- Removed warning messages for out of bounds values for palette and pattern config
- Changed config window size to not have as much unnecessary empty space
- Palette is now plural on config window
- C++20

McAwesome123 committed:
screams in visual studio fuckery

due to some weird fuckery, the c++20 conversion wasn't actually included in the last commit
(ps: explicit casts seem to cause vs to not complain about type changes)

McAwesome123 committed:
now builds as yok.scr

inb4 this change fucks something up

McAwesome123 commented:
6 am programming goes brrrr

pierce-smith1 reviewed (approved):
It's release time motherfucker.
